core/src/main/java/org/pentaho/di/core/hash/LongHashIndex.java

```diff
@@ -0,0 +1,200 @@
+
+    index = new LongHashIndexEntry[factor2Size];
+  }
+
+  /**
+   * Create a new long/long hash index
+   */
+  public LongHashIndex() {
+    this( STANDARD_INDEX_SIZE );
+  }
+
+  public int getSize() {
+    return size;
+  }
+
+  public boolean isEmpty() {
+    return size == 0;
+  }
+
+  public Long get( long key ) throws KettleValueException {
+    int hashCode = generateHashCode( key );
+
+    int indexPointer = indexFor( hashCode, index.length );
+    LongHashIndexEntry check = index[indexPointer];
+
+    while ( check != null ) {
+      if ( check.hashCode == hashCode && check.equalsKey( key ) ) {
+        return check.value;
+      }
+      check = check.nextEntry;
+    }
+    return null;
+  }
+
+  public void put( long key, Long value ) throws KettleValueException {
+    int hashCode = generateHashCode( key );
+    int indexPointer = indexFor( hashCode, index.length );
+
+    LongHashIndexEntry check = index[indexPointer];
+    LongHashIndexEntry previousCheck = null;
+
+    while ( check != null ) {
+
+      // If there is an identical entry in there, we replace the entry
+      // And then we just return...
+      //
+      if ( check.hashCode == hashCode && check.equalsKey( key ) ) {
+        check.value = value;
+        return;
+      }
+      previousCheck = check;
+      check = check.nextEntry;
+    }
+
+    // Don't forget to link to the previous check entry if there was any...
+    //
+    if ( previousCheck != null ) {
+      previousCheck.nextEntry = new LongHashIndexEntry( hashCode, key, value, null );
+    } else {
+      index[indexPointer] = new LongHashIndexEntry( hashCode, key, value, null );
+    }
+
+    // If required, resize the table...
+    //
+    resize();
+  }
+
+  private final void resize() {
+    // Increase the size of the index...
+    //
+    size++;
+
+    // See if we've reached our resize threshold...
+    //
+    if ( size >= resizeThresHold ) {
+
+      LongHashIndexEntry[] oldIndex = index;
+
+      // Double the size to keep the size of the index a factor of 2...
+      // Allocate the new array...
+      //
+      int newSize = 2 * index.length;
+
+      LongHashIndexEntry[] newIndex = new LongHashIndexEntry[newSize];
+
+      // Loop over the old index and re-distribute the entries
+      // We want to make sure that the calculation
+      // entry.hashCode & ( size - 1)
+      // ends up in the right location after re-sizing...
+      //
+      for ( int i = 0; i < oldIndex.length; i++ ) {
+        LongHashIndexEntry entry = oldIndex[i];
+        if ( entry != null ) {
+          oldIndex[i] = null;
+
+          // Make sure we follow all the linked entries...
+          // This is a bit of extra work, TODO: see how we can avoid it!
+          //
+          do {
+            LongHashIndexEntry next = entry.nextEntry;
+            int indexPointer = indexFor( entry.hashCode, newSize );
+            entry.nextEntry = newIndex[indexPointer];
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).